libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx

```diff
@@ -697,7 +1056,146 @@ std::map<Int_t,std::map<Int_t,Int_t>> TAngularCorrelation::GenerateIndexMap(std:
+
+   // get number of group entries
+   Int_t size = groupangles.size();
+
+   // basic consistency check
+   if(size!=(Int_t) groupweights.size()) {
+      printf("Not all groups have been assigned an angle.\n");   
+   } 
+   
+   // initialize group vector
+   for(Int_t i=0;i<size;i++) {
+       groupangle.push_back(groupangles[i]);
+   }
+
+   return groupangle;
+}
+////////////////////////////////////////
+/// Generate Folded Angles
+///gives Cos(RADIAN) angles out for each folded index
+///
+/// \param[in] anglemap (can be for grouped or ungrouped anglular indexes)
+
+
+std::vector<Double_t> TAngularCorrelation::GenerateFoldedAngles(std::vector<Double_t> &anglemap)
+{
+   std::vector<Double_t> folds; // vector to return
+
+   // get size
+   Int_t size = anglemap.size();
+
+   //consistancy check
+   if(size == 0){
+      printf("Angles have not been assigned yet. \n");
+      printf("Therefore cannot fold indexes. \n");
+      return folds;
+      }
+   
+  
+   //declare fold array and fill with zeros
+   Double_t foldArray[size];
+   for(Int_t i=0; i<size; i++){
+      foldArray[i]=0;
+   }
+
+  //fill fold array with angular index and abs(cos(angle)) values
+     for (Int_t i=0; i<size; i++){
+         Double_t angle = anglemap[i];
+         Double_t abs_angle = fabs(cos(angle));
+         foldArray[i] = abs_angle;           
+      }
+  
+   // Iterate through fold array
+   for(Int_t i=0; i<size; i++){
+        Double_t fold_angle = foldArray[i];
+        Bool_t alreadyclaimed = kFALSE;
+
+      for (Int_t m=0; m<folds.size(); m++) {  
+           if(TMath::Abs(fold_angle-folds[m])<0.00005){//look for duplicated angles in the fold array 
+           alreadyclaimed = kTRUE;
+           break;
+          }
+       }
+     if (!alreadyclaimed){
+         folds.push_back(fold_angle); 
+        
+       } 
+     
+            
+  }
+  
+  
+  std::sort(folds.begin(),folds.end()); 
+
+   return folds;
+}
 
+////////////////////////////////////////
+/// Assign Groups for Angular Indexes
+///
+/// \param[in] group vector of assigned groups for angular indexes
+/// \param[in] anglemap
+
+std::vector<Int_t> TAngularCorrelation::GenerateFoldedIndexes(std::vector<Double_t> &folds, std::vector<Double_t> &anglemap)
+{
+   std::vector<Int_t> fold_indexes; // vector to return
+
+  // get sizes of fold and angle vectors
+  Int_t fold_size = folds.size();
+  Int_t angle_size = anglemap.size();
+
+  // itterate through angle map to find abs(cos(angle)
+   for(Int_t i=0;i<angle_size;i++) {
+      Double_t angle = anglemap[i];
+      Double_t cos_angle = fabs(cos(angle));
+    
+      //itterate through folds
+      for(Int_t j=0; j<fold_size; j++){
+         Double_t fold_angle = folds[j];
+         if(TMath::Abs(fold_angle-cos_angle)<0.00005){//compare abs(cos(angle)) to fold_Angles to find matches
+            fold_indexes.push_back(j);//assign folds 
+            }
+         }
+     }
+ 
+
+   return fold_indexes;
+}
```

Comment:
Since GenerateFoldedAngles and GenerateFoldedIndexes are doing very similar things, can we combine the two into a single function?

Also, this needs to be saved to the class.